File: src/tagging/importers/opensubtitles.rs
use lazy_regex::regex;
use reqwest::StatusCode;
use serde::Deserialize;
use serde_json::json;
use std::{cmp::Ordering, time::SystemTime};
use tokio::sync::Mutex;

#[derive(Deserialize)]
struct LoginResponse {
    token: String,
}

pub struct OpenSubtitles {
    agent: reqwest::Client,
    api_key: String,
    username: String,
    password: String,
    auth_token: Mutex<Option<(SystemTime, String)>>,
}
impl OpenSubtitles {
    pub fn new(api_key: String, username: String, password: String) -> Self {
        return Self {
            agent: reqwest::Client::new(),
            api_key,
            username,
            password,
            auth_token: Mutex::new(None),
        };
    }
    async fn login(&self, auth_token: &mut Option<(SystemTime, String)>) -> reqwest::Result<()> {
        let response: LoginResponse = self
            .agent
            .post("https://api.opensubtitles.com/api/v1/login")
            .header("User-Agent", "plex-autotagger")
            .header("Api-Key", &self.api_key)
            .json(&json!({
                "username": &self.username,
                "password": &self.password,
            }))
            .send()
            .await?
            .json()
            .await?;

        *auth_token = Some((SystemTime::now(), response.token));

        return Ok(());
    }

    async fn authenticated(
        &self,
        req_fn: impl Fn() -> reqwest::RequestBuilder,
    ) -> reqwest::Result<reqwest::Response> {
        println!("Making OST request");
        let query_start = SystemTime::now();
        loop {
            let mut auth_token = self.auth_token.lock().await;
            if let Some((ref token_time, ref mut token)) = *auth_token {
                let response = req_fn()
                    .header("User-Agent", "plex-autotagger")
                    .header("Api-Key", &self.api_key)
                    .header("Authorization", String::from("Bearer ") + token)
                    .send()
                    .await?;
                if response.status() == StatusCode::UNAUTHORIZED {
                    if token_time < &query_start {
                        self.login(&mut auth_token).await?;
                        continue;
                    } else {
                        return Err(response.error_for_status().unwrap_err());
                    }
                } else {
                    return Ok(response);
                }
            } else {
                self.login(&mut auth_token).await?;
                continue;
            }
        }
    }

    pub async fn find_subtitles(&self, tmdb_id: i32) -> reqwest::Result<Vec<SubtitleSummary>> {
        let search_result: SearchResults = self
            .authenticated(|| {
                self.agent
                    .get("https://api.opensubtitles.com/api/v1/subtitles")
                    .query(&[("tmdb_id", &tmdb_id.to_string())])
            })
            .await?
            .json()
            .await?;

        let mut files: Vec<SubtitleSummary> = search_result
            .data
            .iter()
            .filter(|subtitle| &subtitle.attributes.language == "en")
            .flat_map(|subtitle| {
                subtitle
                    .attributes
                    .files
                    .iter()
                    .map(|file| SubtitleSummary {
                        name: format!(
                            "lang: {}, name: {}, uploader: {} ({})",
                            subtitle.attributes.language,
                            file.file_name,
                            subtitle.attributes.uploader.name,
                            subtitle.attributes.uploader.rank,
                        ),
                        file_id: file.file_id,
                        uploader: subtitle.attributes.uploader.clone(),
                    })
            })
            .collect();

        files.sort_by(|a, b| {
            match numeric_rank(&a.uploader.rank).cmp(&numeric_rank(&b.uploader.rank)) {
                Ordering::Less => return Ordering::Less,
                Ordering::Greater => return Ordering::Greater,
                Ordering::Equal => return Ordering::Equal,
            }
            // May add more criteria later. Not sure yet.
        });

        return Ok(files);
    }

    pub async fn download_subtitles(&self, file_id: u32) -> reqwest::Result<String> {
        let pointer: DownloadPointer = self
            .authenticated(|| {
                self.agent
                    .post("https://api.opensubtitles.com/api/v1/download")
                    .json(&json!({
                        "file_id": file_id,
                    }))
            })
            .await?
            .json()
            .await?;

        return Ok(self
            .authenticated(|| self.agent.get(&pointer.link))
            .await?
            .text()
            .await?);
    }

    /// Finds the best subtitles by grabbing up to 3 and comparing them. The one that
    /// matches more closely with the rest gets picked.
    /// returns `(file_name, subtitles)`
    pub async fn find_best_subtitles(&self, tmdb_id: i32) -> anyhow::Result<(String, String)> {
        let subtitle_results = self.find_subtitles(tmdb_id).await?.into_iter().take(3);

        let mut subtitles = Vec::with_capacity(3);
        for subtitle_summary in subtitle_results {
            subtitles.push((
                subtitle_summary.name,
                self.download_subtitles(subtitle_summary.file_id).await?,
            ));
        }

        return tokio::task::spawn_blocking(move || match subtitles.len() {
            0 => anyhow::bail!("No subtitles found"),
            1 => return Ok(subtitles.pop().unwrap()),
            2 => {
                let file1 = subtitles.pop().unwrap();
                let file2 = subtitles.pop().unwrap();
                let distance = levenshtein::levenshtein(&file1.1, &file2.1);
                let max_distance = file1.1.len().max(file2.1.len());
                if distance > max_distance / 4 {
                    anyhow::bail!("Couldn't find reliable subtitles");
                }
                return Ok(file1);
            }
            3 => {
                let file1 = subtitles.pop().unwrap();
                let file2 = subtitles.pop().unwrap();
                let file3 = subtitles.pop().unwrap();
                let mut distance1: Option<usize> = None;
                let mut distance2: Option<usize> = None;
                let mut distance3: Option<usize> = None;
                let file1_stripped = strip_subtitles(&file1.1);
                let file2_stripped = strip_subtitles(&file2.1);
                let file3_stripped = strip_subtitles(&file3.1);
                rayon::scope(|s| {
                    s.spawn(|_| {
                        distance1 = Some(levenshtein::levenshtein(&file1_stripped, &file2_stripped))
                    });
                    s.spawn(|_| {
                        distance2 = Some(levenshtein::levenshtein(&file2_stripped, &file3_stripped))
                    });
                    s.spawn(|_| {
                        distance3 = Some(levenshtein::levenshtein(&file3_stripped, &file1_stripped))
                    });
                });
                let distance1 = distance1.unwrap();
                let distance2 = distance2.unwrap();
                let distance3 = distance3.unwrap();
                let max_distance = file1.1.len().max(file2.1.len()).max(file3.1.len());

                let mut distances = vec![(1, distance1), (2, distance2), (3, distance3)];
                distances.sort_by_key(|item| item.1);

                if distances[0].0 > max_distance / 4 {
                    anyhow::bail!("Couldn't find reliable subtitles");
                }

                if distances[0].0 == 1 {
                    if distances[1].0 == 2 {
                        return Ok(file2);
                    } else {
                        return Ok(file1);
                    }
                } else if distances[0].0 == 2 {
                    if distances[1].0 == 1 {
                        return Ok(file2);
                    } else {
                        return Ok(file3);
                    }
                } else if distances[0].0 == 3 {
                    if distances[1].0 == 1 {
                        return Ok(file1);
                    } else {
                        return Ok(file3);
                    }
                }

                unreachable!();
            }
            _ => unreachable!(),
        })
        .await
        .unwrap();
    }
}

/// Strips symbols from subtitles that may cause issues during comparison
pub fn strip_subtitles(subs: &str) -> String {
    let intermediate = regex!(
        r"(?:<\s*[^>]*>|<\s*/\s*a>)|(?:^.*-->.*$|^[0-9]+$|[^a-zA-Z0-9 ?\.,!\n]|^\s*-*\s*|\r)"m
    )
    .replace_all(subs, "");
    return regex!(r"[\n ]{1,}")
        .replace_all(&intermediate, " ")
        .into_owned();
}

/// Converts an uploader's rank into a numeric value for sorting
fn numeric_rank(rank: &str) -> usize {
    return match a.uploader.rank {
        "Administrator" => 100,
        "Application Developers" => 40,
        "Gold member" => 30,
        "Bronze Member" => 20,
        "anonymous" => 0,
        _ => 10,
    };
}

#[derive(Debug, Deserialize, Clone)]
struct SearchResults {
    // total_pages: u32,
    // total_count: u32,
    // per_page: u32,
    // page: u32,
    data: Vec<SearchResult>,
}

#[derive(Debug, Deserialize, Clone)]
struct SearchResult {
    // id: String,
    // #[serde(rename = "type")]
    // result_type: String,
    attributes: STAttributes,
}

#[derive(Debug, Deserialize, Clone)]
struct STAttributes {
    // subtitle_id: String,
    language: String,
    // download_count: u32,
    // new_download_count: u32,
    // hearing_impaired: bool,
    // votes: u32,
    // ratings: f32,
    // from_trusted: bool,
    // foreign_parts_only: bool,
    // ai_translated: bool,
    // machine_translated: bool,
    // release: String,
    uploader: OSTUploader,
    files: Vec<STFile>,
}

#[derive(Debug, Deserialize, Clone)]
struct OSTUploader {
    // uploader_id: i32,
    name: String,
    rank: String,
}

#[derive(Debug, Deserialize, Clone)]
struct STFile {
    file_id: u32,
    file_name: String,
}

#[derive(Debug, Clone)]
pub struct SubtitleSummary {
    name: String,
    file_id: u32,
    uploader: OSTUploader,
}

#[derive(Debug, Deserialize, Clone)]
struct DownloadPointer {
    link: String,
    // file_name: String,
    // requests: u32,
    // remaining: u32,
    // message: String,
    // reset_time: String,
    // reset_time_utc: String,
}

File: backend/helpers/opensubtitles/opensubtitles.go
package opensubtitles

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/agnivade/levenshtein"
	"github.com/sploders101/mediacorral/backend/dbapi"
	"github.com/sploders101/mediacorral/backend/helpers/blobs"

	proto "github.com/sploders101/mediacorral/backend/gen/mediacorral/server/v1"
)

const (
	LOGIN_URL                 string        = "https://api.opensubtitles.com/api/v1/login"
	SEARCH_URL                string        = "https://api.opensubtitles.com/api/v1/subtitles"
	DOWNLOAD_URL              string        = "https://api.opensubtitles.com/api/v1/download"
	DEFAULT_TIMEOUT           time.Duration = time.Minute
	SUBTITLE_COMPARISON_LIMIT int           = 3
)

var (
	STRIP_SUBTITLE_REGEX *regexp.Regexp = regexp.MustCompile(
		`(?m)(?:<\s*[^>]*>|<\s*/\s*a>)|(?:^.*-->.*$|^[0-9]+$|[^a-zA-Z0-9 ?\.,!\n]|^\s*-*\s*|\r)`,
	)
	STRIP_WHITESPACE_REGEX *regexp.Regexp = regexp.MustCompile(`(?m)[\n ]{1,}`)
	ErrNoSubtitles                        = errors.New("no subtitles were found")
	ErrUnreliableSubtitles                = errors.New(
		"subtitles were found, but were inconsistent",
	)
)

type OstImporter struct {
	client http.Client

	apiKey   string
	username string
	password string

	authMutex    sync.Mutex
	authToken    string
	authIssuedAt time.Time
}

func NewOstImporter(apiKey, username, password string) (*OstImporter, error) {
	return &OstImporter{
		client: http.Client{
			Timeout: DEFAULT_TIMEOUT,
		},
		apiKey:   apiKey,
		username: username,
		password: password,
	}, nil
}

// Logs into OST and renews the authToken.
//
// This function should only be called while a mutex lock is held.
func (importer *OstImporter) login() error {
	loginCredsBuf, err := json.Marshal(struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}{
		Username: importer.username,
		Password: importer.password,
	})
	if err != nil {
		return err
	}

	res, err := http.Post(LOGIN_URL, "application/json", bytes.NewBuffer(loginCredsBuf))
	if err != nil {
		return err
	}
	defer func() {
		_ = res.Body.Close()
	}()

	var response struct {
		Token string `json:"token"`
	}
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		return err
	}

	importer.authToken = response.Token
	importer.authIssuedAt = time.Now()

	return nil
}

// Adds authentication details to a request and runs it, handling token
// expiration automatically.
//
// `authMutex` is handled internally, unlike `login`. Calling this function
// with a locked `authMutex` will deadlock.
func (importer *OstImporter) runAuthenticated(
	req *http.Request,
	returnNon200 bool,
) (*http.Response, error) {
	importer.authMutex.Lock()
	if importer.authToken == "" {
		if err := importer.login(); err != nil {
			importer.authMutex.Unlock()
			return nil, err
		}
	}
	lastRefresh := importer.authIssuedAt
	authToken := importer.authToken
	importer.authMutex.Unlock()

	for {
		// Make request
		req.Header.Set("user-agent", "Mediacorral v1.0.0")
		req.Header.Set("api-key", importer.apiKey)
		req.Header.Set("authorization", "Bearer "+authToken)
		res, err := importer.client.Do(req)
		if err != nil {
			return nil, err
		}

		// Validate request
		switch res.StatusCode {
		case 200:
			return res, nil
		case 401:
			// We are unauthorized. Our token must have expired.
			importer.authMutex.Lock()
			if lastRefresh.Compare(importer.authIssuedAt) >= 0 {
				// Token was not renewed asynchronously. Current token is invalid.
				// Refresh it.
				err := importer.login()
				importer.authMutex.Unlock()
				if err != nil {
					return nil, err
				}
			} else {
				// Token was already renewed. Update our local cache & unlock.
				lastRefresh = importer.authIssuedAt
				authToken = importer.authToken
				importer.authMutex.Unlock()
			}
		default:
			if returnNon200 {
				return res, nil
			} else {
				return nil, fmt.Errorf("received status code %d from %s", res.StatusCode, req.URL.String())
			}
		}
	}
}

// Makes a request (with authentication) and automatically decodes the response
func (importer *OstImporter) makeRequest(req *http.Request, respValue any) error {
	resp, err := importer.runAuthenticated(req, false)
	if err != nil {
		return err
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if !slices.Contains(
		[]string{"application/json", "application/json; charset=utf-8"},
		resp.Header.Get("content-type"),
	) {
		bodyBytes, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		slog.Debug(
			"Received non-json response from OST API",
			"url", req.URL.String(),
			"contentType", resp.Header.Get("content-type"),
			"response", base64.RawStdEncoding.EncodeToString(bodyBytes),
		)
		return fmt.Errorf("received non-json response from %s", req.URL.String())
	}
	if err := json.NewDecoder(resp.Body).Decode(respValue); err != nil {
		return err
	}
	return nil
}

func (importer *OstImporter) FindSubtitles(tmdbId int32) ([]SubtitleSummary, error) {
	req, err := http.NewRequest("GET", SEARCH_URL, nil)
	if err != nil {
		return nil, err
	}

	query := req.URL.Query()
	query.Set("tmdb_id", strconv.FormatInt(int64(tmdbId), 10))
	req.URL.RawQuery = query.Encode()

	var results ostSearchResults
	if err := importer.makeRequest(req, &results); err != nil {
		return nil, err
	}

	var files []SubtitleSummary
	for _, subtitle := range results.Data {
		// Filters
		if subtitle.Attributes.Language == nil {
			continue
		}
		if *subtitle.Attributes.Language != "en" {
			continue
		}

		lang := *subtitle.Attributes.Language

		for _, file := range subtitle.Attributes.Files {
			files = append(files, SubtitleSummary{
				Name: fmt.Sprintf(
					"lang: %s, name: %s, uploader: %s (%s)",
					lang,
					file.FileName,
					subtitle.Attributes.Uploader.Name,
					subtitle.Attributes.Uploader.Rank,
				),
				DownloadCount:    subtitle.Attributes.DownloadCount,
				NewDownloadCount: subtitle.Attributes.NewDownloadCount,
				FileId:           file.FileId,
				Uploader:         subtitle.Attributes.Uploader,
			})
		}
	}

	slices.SortFunc(files, func(a, b SubtitleSummary) int {
		switch {
		case a.Uploader.Rank == "Admin Warning":
			return 1
		case b.Uploader.Rank == "Admin Warning":
			return -1
		case a.NewDownloadCount < b.NewDownloadCount:
			return 1
		case a.NewDownloadCount > b.NewDownloadCount:
			return -1
		case a.DownloadCount < b.DownloadCount:
			return 1
		case a.DownloadCount > b.DownloadCount:
			return -1
		}
		aNumRank := numericRank(a.Uploader.Rank)
		bNumRank := numericRank(b.Uploader.Rank)
		return aNumRank - bNumRank
	})

	return files, nil
}

// Downloads a set of subtitles, returning them as a string
func (importer *OstImporter) DownloadSubtitles(fileId uint32) (string, error) {
	bodyBuf, err := json.Marshal(struct {
		FileId uint32 `json:"file_id"`
	}{
		FileId: fileId,
	})
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest("POST", DOWNLOAD_URL, bytes.NewReader(bodyBuf))
	if err != nil {
		return "", err
	}

	var results struct {
		Link string `json:"link"`
	}
	if err := importer.makeRequest(req, &results); err != nil {
		return "", err
	}

	if results.Link == "" {
		return "", errors.New("OST returned an empty link")
	}

	req, err = http.NewRequest("GET", results.Link, nil)
	if err != nil {
		return "", err
	}
	res, err := importer.runAuthenticated(req, false)
	if err != nil {
		return "", err
	}

	subs, err := io.ReadAll(res.Body)
	if err != nil {
		return "", err
	}
	return string(subs), nil
}

// Finds the best subtitles by grabbing up to 3 and comparing them. One of the subtitles
// that match more closely with the others will be picked. This attempts to weed out
// outliers.
func (importer *OstImporter) FindBestSubtitles(tmdbId int32) (FindBestSubtitlesResult, error) {
	// Find subtitles
	subtitleResults, err := importer.FindSubtitles(tmdbId)
	if err != nil {
		return FindBestSubtitlesResult{}, err
	}
	if len(subtitleResults) == 0 {
		return FindBestSubtitlesResult{}, ErrNoSubtitles
	}

	// Download subtitles
	subtitles := make([]FindBestSubtitlesResult, SUBTITLE_COMPARISON_LIMIT)
	for i, subtitle := range subtitleResults {
		if i >= SUBTITLE_COMPARISON_LIMIT {
			break
		}
		subs, err := importer.DownloadSubtitles(subtitle.FileId)
		if err != nil {
			return FindBestSubtitlesResult{}, err
		}
		subtitles = append(subtitles, FindBestSubtitlesResult{
			Filename:          subtitle.Name,
			Subtitles:         subs,
			minifiedSubtitles: StripSubtitles(subs),
		})
	}

	// Compare subtitles
	//
	// This performs cross-analysis by computing the distance of every permutation and
	// collecting all the distances where each result was involved. The distances of
	// each result are then averaged. The average is used to score the subtitles in
	// aggregate. An outlier will have a higher score when matched with the
	// [theoretically] more frequent "correct" subtitles, of which there will be more,
	// resulting in a higher overall average score. Finally, the subtitles are sorted
	// by their score (TODO: come up with a min function to make this more efficient),
	// and the subtitles with the lowest score are chosen.

	type comparison struct {
		aIndex   int
		bIndex   int
		distance int
	}
	var resultChan = make(chan comparison)
	for aIndex, resultA := range subtitles {
		for bIndex, resultB := range subtitles[aIndex+1:] {
			go func() {
				distance := levenshtein.ComputeDistance(
					resultA.minifiedSubtitles,
					resultB.minifiedSubtitles,
				)
				resultChan <- comparison{
					aIndex:   aIndex,
					bIndex:   bIndex,
					distance: distance,
				}
			}()
		}
	}

	distancesByIndex := make(map[int][]int)
	for result := range resultChan {
		// Add the distance to both indexes. The differences will surface when we average them.
		distancesByIndex[result.aIndex] = append(distancesByIndex[result.aIndex], result.distance)
		distancesByIndex[result.bIndex] = append(distancesByIndex[result.bIndex], result.distance)
	}

	type averagedComparison struct {
		index            int
		averagedDistance int
	}
	var averagedResults []averagedComparison
	for resultIndex, distances := range distancesByIndex {
		averagedResults = append(averagedResults, averagedComparison{
			index:            resultIndex,
			averagedDistance: average(distances),
		})
	}

	topAverage := slices.MinFunc(averagedResults, func(a, b averagedComparison) int {
		return a.averagedDistance - b.averagedDistance
	})
	topResult := subtitles[topAverage.averagedDistance]

	// Check that the results are at least somewhat similar
	maxDistance := len(slices.MaxFunc(subtitles, func(a, b FindBestSubtitlesResult) int {
		return len(a.minifiedSubtitles) - len(b.minifiedSubtitles)
	}).minifiedSubtitles)
	if topAverage.averagedDistance > maxDistance/2 {
		return FindBestSubtitlesResult{}, ErrUnreliableSubtitles
	}

	return FindBestSubtitlesResult{
		Filename:  topResult.Filename,
		Subtitles: topResult.Subtitles,
	}, nil
}

func (importer *OstImporter) GetSubtitles(
	dbTx *dbapi.DbTx,
	blobController *blobs.BlobStorageController,
	videoType proto.VideoType,
	videoId int64,
	tmdbId int32,
) (GetSubtitlesResult, error) {
	results, err := dbTx.GetOstDownloadItemsByMatch(videoType, videoId)
	if err != nil {
		return GetSubtitlesResult{}, err
	}

	if len(results) > 0 {
		existingSubs := results[0]
		filePath := blobController.GetFilePath(existingSubs.BlobId)
		subsBuf, err := os.ReadFile(filePath)
		if err != nil {
			return GetSubtitlesResult{}, err
		}
		return GetSubtitlesResult{
			SubtitlesItem: existingSubs,
			Subtitles:     string(subsBuf),
		}, nil
	}

	subsResult, err := importer.FindBestSubtitles(tmdbId)
	if err != nil {
		return GetSubtitlesResult{}, err
	}
	newSubsItem, err := blobController.AddOstSubtitles(
		dbTx,
		videoType,
		videoId,
		subsResult.Filename,
		subsResult.Subtitles,
	)
	if err != nil {
		return GetSubtitlesResult{}, err
	}
	return GetSubtitlesResult{
		SubtitlesItem: newSubsItem,
		Subtitles:     subsResult.Subtitles,
	}, nil
}

// Strips subtitles into a long sequence of words with HTML formatting removed
// (using a naive regex approach), removing extra information not relevant for
// comparison
func StripSubtitles(subsText string) string {
	subsText = STRIP_SUBTITLE_REGEX.ReplaceAllLiteralString(subsText, "")
	return STRIP_WHITESPACE_REGEX.ReplaceAllLiteralString(subsText, " ")
}

// Assigns a numerical value to an uploader rank. Used for sorting files by what
// should hopefully be the reliability of the uploader.
func numericRank(rank string) int {
	switch strings.ToLower(rank) {
	case "administrator":
		return 0
	case "application developers":
		return 10
	case "gold member":
		return 20
	case "bronze member":
		return 30
	case "anonymous":
		return 100
	case "admin warning":
		return 110
	default:
		return 90
	}
}

// This is a little ridiculous. Why exactly is this not built in?
func average(numbers []int) int {
	sum := 0
	for _, num := range numbers {
		sum += num
	}
	return sum / len(numbers)
}
